rat/scanner.py

```diff
@@ -214,6 +228,14 @@ def raise_error(self, msg):
     # The following functions are scanner states
 
     def delimeter_state(self):
+        bracket_dict = {")": "(", "}": "{", "]": "["}
+        if self.current_char in bracket_dict.values():
+            self.bracket_stack.append(self.current_char)
+        elif self.current_char in bracket_dict.keys():
+            if len(self.bracket_stack) == 0:
+                self.raise_error(f"Found unmatching brackets {self.current_char}")
+            elif self.bracket_stack.pop() != bracket_dict[self.current_char]:
+                self.raise_error(f"Found unmatching brackets {self.current_char}")
```

Comment:
Is the bracket_stack for better errors or is it also doing something else and the errors are convenient?

Either way exciting to have errors like this!

Reply:
It's for error checking. The bracket checking stuff was in the parser, but it turns out the scanner does a single pass of the entire model string so the bracket checking is free if done on the parser.